# <span style="color:blue">Introduction to Pandas</span>

## Importing Library

In [1]:
import pandas as pd

# • Series

## What is a Series
A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [2]:
a = [1,7,2]
var = pd.Series(a)
print(var)

0    1
1    7
2    2
dtype: int64


## Labels
If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [3]:
print(a[0])

1


## Create Labels
With the `index` argument, you can name your own labels.

In [4]:
b = pd.Series(a, index = ["x","y","z"])
print(b)

x    1
y    7
z    2
dtype: int64


When you have created labels, you can access an item by referring to the label.

In [5]:
print(b["y"])

7


## Key/Value Objects as Series
You can also use a key/value object, like a dictionary, when creating a Series.<br>
`Note : The Keys of the dictionary become the lables. `

In [6]:
calories = {"day1":420, "day2":380, "day3":390}
c = pd.Series(calories)
print(c)

day1    420
day2    380
day3    390
dtype: int64


To select only some of the items in the dictionary, use the `index` argument and specify only the items you want to include in the Series.

In [7]:
calories = {"day1":420,"day2":380,"day3":390}
c = pd.Series(calories, index = ["day1", "day2"])
print(c)

day1    420
day2    380
dtype: int64


 # • DataFrames

## What is a DataFrame?
A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [9]:
data = {
    "calories":[420, 380, 390],
    "duration":[50, 40, 45]
}
df = pd.DataFrame(data)
print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


## Locate Row
As you can see from the result above, the DataFrame is like a table with rows and columns.

Pandas use the `loc` attribute to return one or more specified row(s)

In [10]:
print(df.loc[0])

calories    420
duration     50
Name: 0, dtype: int64


`Note: This example returns a Pandas Series.`

In [11]:
#use a list of indexes
print(df.loc[[0,1]])

   calories  duration
0       420        50
1       380        40


`Note: When using [], the result is a Pandas DataFrame.`

## Named Indexes
With the `index` argument, you can name your own indexes.

In [12]:
data = {
    "calories":[420, 380, 390],
    "duration":[50, 40, 45]
}
df = pd.DataFrame(data, index = ["day1", "day2", "day3"])
print(df)

      calories  duration
day1       420        50
day2       380        40
day3       390        45


## Locate Named Indexes
Use the named index in the `loc` attribute to return the specified row(s)

In [13]:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


# • Read CSV

A simple way to store big data sets is to use CSV files (comma separated files).

CSV files contains plain text and is a well know format that can be read by everyone including Pandas.

In [14]:
df = pd.read_csv('data.csv')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

Tip: use `to_string()` to print the entire DataFrame.

## By default, when you print a DataFrame, you will only get the first 5 rows, and the last 5 rows:

In [15]:
print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


# • Read JSON

Big data sets are often stored, or extracted as JSON.

JSON is plain text, but has the format of an object, and is well known in the world of programming, including Pandas.

 ## Load the JSON file into a DataFrame

In [16]:
df = pd.read_json('data.json')
print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

Tip: use `to_string()` to print the entire DataFrame.

## Dictionary as JSON

`JSON = Python Dictionary`

`JSON objects have the same format as Python dictionaries.`

### If your JSON code is not in a file, but in a Python Dictionary, you can load it into a DataFrame directly:

In [18]:
data = data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}
df = pd.DataFrame(data)
print(df)

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


# • Analyzing DataFrames

## Viewing the Data

One of the most used method for getting a quick overview of the DataFrame, is the `head()` method.

The `head()` method returns the headers and a specified number of rows, starting from the top.

`Get a quick overview by printing the first 10 rows of the DataFrame:`

In [19]:
df = pd.read_csv('data.csv')
print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


` Note: if the number of rows is not specified, the head() method will return the top 5 rows. `

In [20]:
print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


There is also a `tail()` method for viewing the last rows of the DataFrame.

The `tail()` method returns the headers and a specified number of rows, starting from the bottom.

`Print the last 5 rows of the DataFrame:`

In [21]:
print(df.tail())

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


## Info About the Data

The DataFrames object has a method called `info()`, that gives you more information about the data set.

In [22]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


## Null Values

The `info()` method also tells us how many Non-Null values there are present in each column, and in our data set it seems like there are 164 of 169 Non-Null values in the "Calories" column.

Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty values. This is a step towards what is called cleaning data, and you will learn more about that in the next chapters.